dask/array/reblock.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import, division, print_function
+from itertools import count, product, chain, accumulate
+from operator import getitem, add, mul, ifloordiv
+import numpy as np
+from toolz import merge, reduce
+from dask.utils import ignoring
+from dask.array.core import rec_concatenate, Array
+
+reblock_names  = ('reblock-%d' % i for i in count(1))
+
+def intersect_blockdims_1d(old, new):
+	summ = sum(old)
+	if summ != sum(new):
+		raise ValueError('sum(old) != sum(new)')
+	od = enumerate(old)
+	nw = iter(new)
+	ret = []
+	sum_old = 0
+	sum_new = 0
+	while True:
+		while sum_new <= sum_old:
+			try:
+				n = next(nw)
+				if not ret or ret[-1]:
+					ret.append([])
+			except StopIteration:
+				break
+			sum_new += n
+			if sum_new - n < sum_old and sum_new >= sum_old:
+				start = o -leftover
+				endd = o
+				adding = (cnt, slice(start, endd))
+
+				ret[-1].append(adding)
+		leftover = 0
+		while sum_old < sum_new:
+			try:
+				cnt, o = next(od)
+			except StopIteration:
+				break
+			sum_old += o
+			if sum_old > sum_new and sum_old - o < sum_new:
+
+				leftover =  (sum_old - sum_new)
+				endd = o - leftover
+				start = 0
+			elif sum_old <= sum_new:
+
+				start = leftover
+				endd = o
+			elif sum_old < sum_new:
+
+				start = 0
+				endd = o
+			else:
+				continue
+
+			adding = (cnt, slice(start, endd))
+			ret[-1].append(adding)
+		if sum_old == summ and sum_new == summ:
+			break
+	return tuple(map(tuple, ret))
+
+def intersect_blockdims(old, new):
+	old_to_new =  tuple(intersect_blockdims_1d(o, n) for o,n in zip(old,new))
+	cross = tuple(product(*old_to_new))
+	cross = tuple(chain(tuple(product(*cr)) for cr in cross))
+	return (old_to_new, cross)
+
+
+def reblock(dsk, blockdims):
+	"""
+	Convert blocks in dsk for new blockdims.
+	"""
```

Comment:
Because this function is user-facing it should have a more extensive docstring and doctest.  Also, you might want to optionally support either the blockdims or blockshape kwargs.  A simple example of a function that supports both is https://github.com/ContinuumIO/dask/blob/master/dask/array/core.py#L28-L44

The first argument, dsk, should probably be changed to `x`.  I tend to use `dsk` as a pronoun for dictionaries that happen to be dask graphs and `x` for full Array objects.